### DISCUSSION TITLE
Error in training 4th FC

### DISCUSSION THREAD

**Discussion by gangliu-github** [ttadano/alamode#126] [Q&A] [answered]
I want to obtain 4th FC for Si for a test, as its simple structure.
However, error occured when training 4th FC. And "SPARSE = 1" has been set already.
Error information:

terminate called recursively
terminate called recursively
terminate called recursively
terminate called recursively
terminate called after throwing an instance of 'terminate called recursively
terminate called recursively
/var/spool/torque/mom_priv/jobs/142574.localhost.SC: line 58: 113194 Aborted                 (core dumped) alm alm_scph.in3 > alm_scph.log3

My input file:

&general
PREFIX = g_quartic
MODE = optimize
NAT = 54; NKD = 1
KD =  C
PERIODIC = 1 1 1;TOLERANCE = 0.001
/
&optimize
  SPARSE = 1
  LMODEL = adaptive-lasso
  DFSET = DFSET_4th
  FC2XML = g.xml
/
&interaction
NORDER = 3  # 1: harmonic, 2: cubic, ..
NBODY = 2 3 3
/

&cell
1.8896
    0.0000000000000000    5.3602212917605723    5.3602212917605723
    5.3602212917605723    0.0000000000000000    5.3602212917605723
    5.3602212917605723    5.3602212917605723    0.0000000000000000
/

&cutoff
*-* None 10 8
/

&position
1    0.0000000000000000    0.0000000000000000    0.0000000000000000
1    0.3333333333333333    0.0000000000000000    0.0000000000000000
1    0.6666666666666665    0.0000000000000000    0.0000000000000000
1    0.0000000000000000    0.3333333333333333    0.0000000000000000
1    0.3333333333333333    0.3333333333333333    0.0000000000000000
1    0.6666666666666666    0.3333333333333334    0.9999999999999999
1    0.0000000000000000    0.6666666666666665    0.0000000000000000
1    0.3333333333333334    0.6666666666666666    0.9999999999999999
1    0.6666666666666666    0.6666666666666666    0.9999999999999999
1    0.0000000000000000    0.0000000000000000    0.3333333333333333
1    0.3333333333333333    0.0000000000000000    0.3333333333333333
1    0.6666666666666667    0.0000000000000000    0.3333333333333333
1    0.0000000000000000    0.3333333333333333    0.3333333333333333
1    0.3333333333333333    0.3333333333333333    0.3333333333333333
1    0.6666666666666667    0.3333333333333333    0.3333333333333333
1    0.0000000000000000    0.6666666666666667    0.3333333333333333
1    0.3333333333333333    0.6666666666666667    0.3333333333333333
1    0.6666666666666667    0.6666666666666667    0.3333333333333333
1    0.0000000000000000    0.0000000000000000    0.6666666666666666
1    0.3333333333333334    0.0000000000000000    0.6666666666666666
1    0.6666666666666665    0.0000000000000000    0.6666666666666666
1    0.0000000000000000    0.3333333333333334    0.6666666666666666
1    0.3333333333333334    0.3333333333333334    0.6666666666666666
1    0.6666666666666666    0.3333333333333334    0.6666666666666665
1    0.0000000000000000    0.6666666666666665    0.6666666666666666
1    0.3333333333333334    0.6666666666666666    0.6666666666666665
1    0.6666666666666666    0.6666666666666666    0.6666666666666665
1    0.0833333333333333    0.0833333333333333    0.0833333333333333
1    0.4166666666666666    0.0833333333333333    0.0833333333333333
1    0.7499999999999999    0.0833333333333334    0.0833333333333332
1    0.0833333333333333    0.4166666666666666    0.0833333333333333
1    0.4166666666666667    0.4166666666666667    0.0833333333333332
1    0.7500000000000000    0.4166666666666667    0.0833333333333331
1    0.0833333333333334    0.7499999999999999    0.0833333333333332
1    0.4166666666666667    0.7500000000000000    0.0833333333333331
1    0.7500000000000000    0.7500000000000000    0.0833333333333331
1    0.0833333333333333    0.0833333333333333    0.4166666666666666
1    0.4166666666666666    0.0833333333333333    0.4166666666666666
1    0.7500000000000002    0.0833333333333335    0.4166666666666665
1    0.0833333333333333    0.4166666666666666    0.4166666666666666
1    0.4166666666666667    0.4166666666666667    0.4166666666666665
1    0.7500000000000002    0.4166666666666667    0.4166666666666664
1    0.0833333333333335    0.7500000000000002    0.4166666666666665
1    0.4166666666666667    0.7500000000000002    0.4166666666666664
1    0.7499999999999999    0.7499999999999999    0.4166666666666666
1    0.0833333333333335    0.0833333333333335    0.7499999999999998
1    0.4166666666666669    0.0833333333333335    0.7499999999999998
1    0.7500000000000002    0.0833333333333336    0.7499999999999997
1    0.0833333333333335    0.4166666666666669    0.7499999999999998
1    0.4166666666666667    0.4166666666666667    0.7500000000000000
1    0.7500000000000002    0.4166666666666670    0.7499999999999997
1    0.0833333333333336    0.7500000000000002    0.7499999999999997
1    0.4166666666666670    0.7500000000000002    0.7499999999999997
1    0.7500000000000002    0.7500000000000002    0.7499999999999997
/

**Comment by ttadano** [comment, marked as answer]
Please give us more detailed information, including the version of ALAMODE, the number of force constants, the number of the training dataset, etc. 

BTW, `SPARSE=1` is only for the ordinary least squares (`LMODEL = ols`) and it does not have any effect for (ada)LASSO .

**Reply by gangliu-github** [reply]
Thanks a lot.

Version: 1.4.2
Training dataset: 199
Atoms in the supercell: 54
  Number of  HARMONIC FCs : 61
  Number of   ANHARM3 FCs : 2284
  Number of   ANHARM4 FCs : 4519

The whole log file is pastaed as following:

 +-----------------------------------------------------------------+
 +                         Program ALM                             +
 +                             Ver.  1.4.2                         +
 +-----------------------------------------------------------------+

 Number of OpenMP threads = 36

 Job started at Thu Feb  2 07:55:04 2023


 Input variables:
 -------------------------------------------------------------------
 General:
  PREFIX = g_quartic
  MODE = optimize
  NAT = 54; NKD = 1
  PRINTSYM = 0; TOLERANCE = 0.001
  KD =    C
  FCSYM_BASIS = Lattice
  PERIODIC =   1  1  1
  MAGMOM = 
  FCS_ALAMODE = 1;  NMAXSAVE = 5
  FC3_SHENGBTE = 0
  FC2_QEFC = 0
  HESSIAN = 0
  FC_ZERO_THR = 0

 Interaction:
  NORDER = 3
  NBODY =   2  3  3

 Optimize:
  LMODEL = adaptive-lasso
  DFSET = DFSET_4th
  NDATA = 199; NSTART = 1; NEND = 199   SKIP = 

  ICONST = 11
  ROTAXIS = 
  FC2XML = g.xml
  FC3XML = 

  SPARSE = 1
  SPARSESOLVER = SimplicialLDLT
  CONV_TOL = 1e-08
  MAXITER = 10000
  MIRROR_IMAGE_CONV = 0

 -------------------------------------------------------------------

 SYSTEM
 ======

  Lattice Vector
    0.000000e+00   1.012867e+01   1.012867e+01 : a1
    1.012867e+01   0.000000e+00   1.012867e+01 : a2
    1.012867e+01   1.012867e+01   0.000000e+00 : a3

  Cell volume = 2.078202e+03

  Reciprocal Lattice Vector
   -3.101682e-01   3.101682e-01   3.101682e-01 : b1
    3.101682e-01  -3.101682e-01   3.101682e-01 : b2
    3.101682e-01   3.101682e-01  -3.101682e-01 : b3

  Atomic species:
     1    C

  Atomic positions in fractional basis and atomic species
     1   0.000000e+00   0.000000e+00   0.000000e+00    1
     2   3.333333e-01   0.000000e+00   0.000000e+00    1
     3   6.666667e-01   0.000000e+00   0.000000e+00    1
     4   0.000000e+00   3.333333e-01   0.000000e+00    1
     5   3.333333e-01   3.333333e-01   0.000000e+00    1
     6   6.666667e-01   3.333333e-01   1.000000e+00    1
     7   0.000000e+00   6.666667e-01   0.000000e+00    1
     8   3.333333e-01   6.666667e-01   1.000000e+00    1
     9   6.666667e-01   6.666667e-01   1.000000e+00    1
    10   0.000000e+00   0.000000e+00   3.333333e-01    1
    11   3.333333e-01   0.000000e+00   3.333333e-01    1
    12   6.666667e-01   0.000000e+00   3.333333e-01    1
    13   0.000000e+00   3.333333e-01   3.333333e-01    1
    14   3.333333e-01   3.333333e-01   3.333333e-01    1
    15   6.666667e-01   3.333333e-01   3.333333e-01    1
    16   0.000000e+00   6.666667e-01   3.333333e-01    1
    17   3.333333e-01   6.666667e-01   3.333333e-01    1
    18   6.666667e-01   6.666667e-01   3.333333e-01    1
    19   0.000000e+00   0.000000e+00   6.666667e-01    1
    20   3.333333e-01   0.000000e+00   6.666667e-01    1
    21   6.666667e-01   0.000000e+00   6.666667e-01    1
    22   0.000000e+00   3.333333e-01   6.666667e-01    1
    23   3.333333e-01   3.333333e-01   6.666667e-01    1
    24   6.666667e-01   3.333333e-01   6.666667e-01    1
    25   0.000000e+00   6.666667e-01   6.666667e-01    1
    26   3.333333e-01   6.666667e-01   6.666667e-01    1
    27   6.666667e-01   6.666667e-01   6.666667e-01    1
    28   8.333333e-02   8.333333e-02   8.333333e-02    1
    29   4.166667e-01   8.333333e-02   8.333333e-02    1
    30   7.500000e-01   8.333333e-02   8.333333e-02    1
    31   8.333333e-02   4.166667e-01   8.333333e-02    1
    32   4.166667e-01   4.166667e-01   8.333333e-02    1
    33   7.500000e-01   4.166667e-01   8.333333e-02    1
    34   8.333333e-02   7.500000e-01   8.333333e-02    1
    35   4.166667e-01   7.500000e-01   8.333333e-02    1
    36   7.500000e-01   7.500000e-01   8.333333e-02    1
    37   8.333333e-02   8.333333e-02   4.166667e-01    1
    38   4.166667e-01   8.333333e-02   4.166667e-01    1
    39   7.500000e-01   8.333333e-02   4.166667e-01    1
    40   8.333333e-02   4.166667e-01   4.166667e-01    1
    41   4.166667e-01   4.166667e-01   4.166667e-01    1
    42   7.500000e-01   4.166667e-01   4.166667e-01    1
    43   8.333333e-02   7.500000e-01   4.166667e-01    1
    44   4.166667e-01   7.500000e-01   4.166667e-01    1
    45   7.500000e-01   7.500000e-01   4.166667e-01    1
    46   8.333333e-02   8.333333e-02   7.500000e-01    1
    47   4.166667e-01   8.333333e-02   7.500000e-01    1
    48   7.500000e-01   8.333333e-02   7.500000e-01    1
    49   8.333333e-02   4.166667e-01   7.500000e-01    1
    50   4.166667e-01   4.166667e-01   7.500000e-01    1
    51   7.500000e-01   4.166667e-01   7.500000e-01    1
    52   8.333333e-02   7.500000e-01   7.500000e-01    1
    53   4.166667e-01   7.500000e-01   7.500000e-01    1
    54   7.500000e-01   7.500000e-01   7.500000e-01    1


  Time Elapsed: 0.037137 sec.

 -------------------------------------------------------------------

 SYMMETRY
 ========

  Space group: Fd-3m (227)
  Number of symmetry operations = 1296

  Given system is not a primitive cell.
  There are    27 translation operations.
  Primitive cell contains 2 atoms

  **Cell-Atom Correspondens Below**
  CELL |  ATOM
     1 |     1   28
     2 |    10   37
     3 |    19   46
     4 |     4   31
     5 |    13   40
     6 |    22   49
     7 |     7   34
     8 |    16   43
     9 |    25   52
    10 |     2   29
    11 |    11   38
    12 |    20   47
    13 |     5   32
    14 |    14   41
    15 |    23   50
    16 |     8   35
    17 |    17   44
    18 |    26   53
    19 |     3   30
    20 |    12   39
    21 |    21   48
    22 |     6   33
    23 |    15   42
    24 |    24   51
    25 |     9   36
    26 |    18   45
    27 |    27   54

  Time Elapsed: 0.132332 sec.

 -------------------------------------------------------------------

 INTERACTION
 ===========

  +++ Cutoff Radii Matrix (NKD x NKD matrix) +++
   HARMONIC
     None

    ANHARM3
       10

    ANHARM4
        8


  List of neighboring atoms below.
  Format [N th-nearest shell, distance (Number of atoms on the shell)]

    1 (  C):   1    2.9239 (  4) -  28(  C)  30(  C)  34(  C)  46(  C)
               2    4.7747 ( 12) -   2(  C)   3(  C)   4(  C)   6(  C)
                                     7(  C)   8(  C)  10(  C)  12(  C)
                                    16(  C)  19(  C)  20(  C)  22(  C)
               3   5.59884 ( 12) -  33(  C)  35(  C)  36(  C)  39(  C)
                                    43(  C)  45(  C)  47(  C)  48(  C)
                                    49(  C)  51(  C)  52(  C)  53(  C)
               4   6.75245 (  6) -  15(  C)  17(  C)  18(  C)  23(  C)
                                    24(  C)  26(  C)
               5   7.35831 (  6) -  29(  C)  31(  C)  32(  C)  37(  C)
                                    38(  C)  40(  C)
               6   8.27003 (  8) -   5(  C)   9(  C)  11(  C)  13(  C)
                                    14(  C)  21(  C)  25(  C)  27(  C)
               7   8.77169 (  5) -  41(  C)  42(  C)  44(  C)  50(  C)
                                    54(  C)

   28 (  C):   1    2.9239 (  4) -   1(  C)   2(  C)   4(  C)  10(  C)
               2    4.7747 ( 12) -  29(  C)  30(  C)  31(  C)  33(  C)
                                    34(  C)  35(  C)  37(  C)  39(  C)
                                    43(  C)  46(  C)  47(  C)  49(  C)
               3   5.59884 ( 12) -   5(  C)   6(  C)   8(  C)  11(  C)
                                    12(  C)  13(  C)  15(  C)  16(  C)
                                    17(  C)  20(  C)  22(  C)  23(  C)
               4   6.75245 (  6) -  42(  C)  44(  C)  45(  C)  50(  C)
                                    51(  C)  53(  C)
               5   7.35831 (  6) -   3(  C)   7(  C)   9(  C)  19(  C)
                                    21(  C)  25(  C)
               6   8.27003 (  8) -  32(  C)  36(  C)  38(  C)  40(  C)
                                    41(  C)  48(  C)  52(  C)  54(  C)
               7   8.77169 (  5) -  14(  C)  18(  C)  24(  C)  26(  C)
                                    27(  C)

  For  ANHARM4, interactions related to more than 3 atoms will be neglected.
  Time Elapsed: 0.313815 sec.

 -------------------------------------------------------------------

 FORCE CONSTANT
 ==============

  Symmetry is handled in crystallographic (fractional) coordinates.

  Number of  HARMONIC FCs : 61
  Number of   ANHARM3 FCs : 2284
  Number of   ANHARM4 FCs : 4519

  Time Elapsed: 2.17852 sec.

 -------------------------------------------------------------------

 CONSTRAINT
 ==========

  ICONST = 1: Constraints for translational invariance
              will be considered.

  FC2XML is given : Harmonic force constants will be 
                    fixed to the values given in g.xml

  Generating constraints from crystal symmetry
  in crystallographic (fractional) coordinates ...
   HARMONIC done.
    ANHARM3 done.
    ANHARM4 done.
  Finished !

  Generating constraints for translational invariance ...
   HARMONIC ... done.
    ANHARM3 ... done.
    ANHARM4 ... done.
  Finished !

  Number of constraints [T-inv, R-inv (self), R-inv (cross)]:
   HARMONIC      2    0    0
    ANHARM3    101    0    0
    ANHARM4   2346    0    0

  There are constraints from crystal symmetry.
  The number of such constraints for each order:
   HARMONIC     40
    ANHARM3   1703
    ANHARM4   3356

  Constraints of T-inv, R-inv (self), and those from crystal symmetry are merged.
  If there are redundant constraints, they are removed in this process.

  Number of inequivalent constraints (self, cross) : 
   HARMONIC     41    0
    ANHARM3   1727    0
    ANHARM4   3961    0

  ICONST >= 10 : Constraints will be considered algebraically.

  Number of free HARMONIC FCs : 0
  Number of free  ANHARM3 FCs : 557
  Number of free  ANHARM4 FCs : 559

  Time Elapsed: 894.078 sec.

 -------------------------------------------------------------------

 OPTIMIZATION
 ============

  LMODEL = adaptive-lasso

  Training data file (DFSET) : DFSET_4th

  NSTART = 1; NEND = 199
  199 entries will be used for training.

  Total Number of Parameters : 6864
  Total Number of Free Parameters : 1116


 WARNING in optimize_main  MESSAGE: STANDARDIZE = 1 in adaptive LASSO is meaningless.
 Switch to STANDARDIZE = 0.

**Reply by ttadano** [reply]
Thanks. I could not find obvious mistakes in the input file. I'm not sure what causes the exception. Can you try again with the following changes?

- a smaller number of training data (set NDATA=10, for example)
- a smaller number of OpenMP thread

**Reply by gangliu-github** [reply]
I tried as you suggest last night.
Number of OpenMP threads = 1 and NDATA = 10 are set. However, the same error also occured.

**Reply by ttadano** [reply]
Can you send me all the input files necessary to reproduce the error via email?

**Reply by gangliu-github** [reply]
OK, I will email the files to terumasa.tadano@gmail.com.
However, I find reduce 4th cutoff may solve the problem. I have created 30 random displaced configures, and reduce the 4th cutoff to 6 bohr (2nd and 3rd cutoffs remain unchanged). Now the optimization is accomplished successfully, and I have gotten the 4th FC. The number of force constants are:
  Number of  HARMONIC FCs : 61
  Number of   ANHARM3 FCs : 2284
  Number of   ANHARM4 FCs : 1799
Then what is the appropriate value of 4th cutoff? Is 6 bohr enough for Si for 4th FC?

**Reply by ttadano** [reply]
Thanks. 

I've examined the issue and found that the error is related to the accumulation of numerical noise when constructing the constraint matrix. 

The error could be suppressed by setting `TOL_CONST = 1.0e-5` in the `&general` field.

However, in this particular case, it is much faster to change the symmetrization basis to Cartesian by setting `FCSYM_BASIS = C` in the `&general` field (no need to specify `TOL_CONST` in this case). With this option, the calculation should become stable no matter what value is used for the cutoff radii.

For bulk diamond, including anharmonic terms up to the 3rd nearest neighbor should be sufficient.

**Reply by gangliu-github** [reply]
Thanks a lot! I set FCSYM_BASIS = C, and it really works! And it is accomplished very fast. Can it work in general case?  Or usually setting TOL_CONST = 1.0e-5 is a general way.

**Reply by ttadano** [reply]
I expect the default choice (`FCSYM_BASIS = L`) should work more robustly in most cases, so the observed issue in the diamond is unexpected. I will think about a potential solution to this stability problem. Anyways, adjusting `TOL_CONST` manually is not recommended for general users.

**Reply by gangliu-github** [reply]
Thanks for your patient reply!
